### ISSUE TITLE
Provide anonymized dataset

### ISSUE THREAD

**Issue by chadwhitacre** [gratipay/inside.gratipay.com#1227]
@techtonik has asked (not finding issue/comment right now) for a public anonymized dataset of the Gittip/Gratipay giving graph, to memorialize the project through visualizations.

@techtonik What info should we include?

```
date,amount,from-uuid,to-uuid
```

Something like that?

**Comment by techtonik** [comment]
**TL;DR**

    weekno, date, amount, nick-from, nick-to

Is what makes this dataset valuable for community.

---

> @techtonik has asked (not finding issue/comment right now) for a public anonymized dataset of the Gittip/Gratipay giving graph, to memorialize the project through visualizations.

It was in a reply to private mail about details of #1205 to me. =)

> @techtonik What info should we include?
>
>        date,amount,from-uuid,to-uuid

The most value for the open source community from this dataset is to see support for different projects over the time. There are many projects that provide anonymity, but Gratipay had unprecedented transparency and openness. I don't think we will ever get a chance to see the dataset like that. Big giants will gather the data anyway, payment companies and banks all have it, but not the community, so my wish is to try our best to like it as public as possible without hurting people feelings.

Why I think we can do this? I don't remember that there was an option to hide earnings and people can see that their in and out flows are public. So let me assume that it served as a strong filter for those who was not comfortable with that. Those who opted in always had a choice not to reveal their identities beyond their pubic Gratipay nicknames and details that they added into (again) public profiles. If there was an option to hide some bits of this info, then we should respect this setting and "anonymize" those balances. It we have time and energy, we can actually ask people who hidden - maybe there is nothing secret left there over these years? Of course, it is not about giving our real names or credit card details. Only nicknames. Maybe handles from Twitter and OSM as well. Did Gratipay allowed to donate to GitHub or Twitter account directly, or they were still redirected to GP handle? I lost my memory. =)

I now realize that what we did is open source support graph - version control system for open financing and gratitude. The whole purpose of this dataset is to help people to build beautiful tools like https://gource.io/ to give credit where credit is due and show that even small donations are a fun an interesting instrument that can bring attention and more serous help to solve real problems.

Now for the technical side. If we agree to preserve nicknames then `date, amount, from, to` should be already good, but I want to also calculate this is relative values. How much `amount` is as a percentage of `total` that was sent for this week payday. I can infer it myself, but to make it easier to digest, we can add  `tick` - relative number of the week when amounts were actually moved. Another missing info -operations when funds are moved in and out of the system - they are not important for me, but as a general curiosity to compare that to Bitcoin's `coinbase` transactions they would be nice (when new money where "minted" in Gratipay system, who provided the "blood" for the system and who were the ones who finally needed it).

I think that's it - `nicknames/project, tickno, date, from, to`. I guess incoming balances were asynchronous, so for them the tickno should be `None`, but date should be correct. From is likely to be `None` as well, because the way how people fill up the system was never public.

I think I should finish now. :D

**Comment by chadwhitacre** [comment]
We were definitely anonymous in the particulars, we can't give nicks with amounts to and from. 😞

**Comment by techtonik** [comment]
Can we ask people to opt-in opt-out for publishing their:

1. nicks (yes/no)
2. weekly totals of to (yes/no)
3. weekly totals of from (yes/no)
4. weekly details of from (yes/no)
5. weekly details of to (yes/no)

If source and target yes are received, we can publish the graph edge in the resulting dataset.

**Comment by chadwhitacre** [comment]
We have no good way to contact everyone, and based on past experience most people wouldn't reply. It wouldn't give us a good enough data set to be interesting. :(

**Comment by chadwhitacre** [comment]
I think we have to let this one go. 😞

**Comment by chadwhitacre** [comment]
Alright, @techtonik, maybe you were right after all. Maybe a suitably anonymized dataset _is_ interesting. :o)

I have received a request from @johanneswachs to provide an anonymized dataset for study:

> My name is Johannes Wachs and I am a postdoc at RWTH Aachen University in Germany. I am writing with a question about the possibility of analyzing legacy data from Gratipay.
>
> Just a bit about me: I research social aspects of the sustainability of open source software ecosystems. One major source of inspiration is Nadia Eghbal's [Roads and Bridges](https://www.fordfoundation.org/about/library/reports-and-studies/roads-and-bridges-the-unseen-labor-behind-our-digital-infrastructure/) white paper. I am fascinated by the unique mechanisms that drive participation, success, and sustainability in open source. I also think that more research is needed to better understand the stresses on these important public goods and how the people behind them can be better supported. Lately I have been thinking about how open source is funded, reading about channels including bug bounties, crowdfunding, corporate sponsorship, and Feross Aboukhadijeh's recent experiment advertising in the terminal. I also read about your experience with Gratipay. 
> 
> And that is why I am writing you: to ask if there is any way my colleagues and I could access the data on contributions. As researchers, we are interested in who contributes to which projects, and what impact contributions have on open source development and sustainability. Are you interested in facilitating such a project and do you think it is feasible to share this data with us? I am happy to draft a more concrete proposal outlining our questions.
> 
> Let me note that the members of our group here at RWTH have extensive experience in handling sensitive data under strict supervision, including an ongoing project with [a high-profile organization] where we analyse [their data] under an NDA. Our work is published in high profile academic venues like TheWebConf, CSCW, HCI, ICWSM, KDD, Nature Human Behavior, Empirical Software Engineering, ACM Trans. On Social Computing, and others.

----

>> I think we might be able to share an anonymized dataset. Can you let me know what questions you are hoping to answer?
>>
>> Can you also let me know your GitHub username? We will need to make our conversation public there in order to proceed.


----

> Our goal is to understand how receiving funds changes the behavior of developers and their collaborators. Here are some example research questions:
>
> 1. Do developers contribute more frequently when they get funded? 
> 1. Do they focus more effort on specific projects? 
> 1. Do they shift their work from nights and weekends to more traditional working hours?
> 1. Are there network effects? Do collaborators of funded developers also change their behavior?
>
> We are excited to find answers to these questions because we think there are big sustainability risks in the open source system and want to understand how best to help. We were happy that Github added a sponsorship feature but feel like research could inform funders how to contribute effectively (i.e. 5$ a month to your 5 favorite developers or 25$ a month to one). I also hope that sharing positive results about outcomes will encourage more people to "tip" the OSS devs they rely on.
>
> To answer these questions it would be great to have a contribution-level dataset: Github User A received X$ at time T. No need for any information about the contributors. Because we are so interested in the behavior of the specific individuals receiving funds, it would be difficult to answer our questions without linking the data to Github accounts. We are happy to sign NDAs and keep the data private if that helps.

Putting this out here as it has been sitting in my inbox for weeks. I will need to find time to work on this, if there are no objections from anyone still listening ...

**Comment by Changaco** [comment]
The weekly income was public data for the vast majority of recipients, so I see no problem with providing it to researchers. In fact we could send them Liberapay's data as well.

**Comment by johanneswachs** [comment]
I appreciate you following up on this! Please let me know if I can provide more information about my plans here. I'm also happy to take requests to visualize the data i.e. as a network.

**Comment by chadwhitacre** [comment]
> Github User A received X$ at time T

This seems to be the heart of the ask. I’m inclined to think a little more broadly and to provide two CSVs:

- `participants.csv`&mdash;ID and username along with accounts elsewhere (Twitter, GitHub, etc.)
- `tips.csv`&mdash;ID, dollar amount, and date

I believe this would encompass your ask, @johanneswachs, while also enabling others who might be interested in the data that Gratipay has available that was public while we were running. Yes?

**Comment by johanneswachs** [comment]
Additional information on the participants would be very useful. These two CSVs cover what I have requested.

One interesting extension would be data on contributors via an anonymized persistent ID for each contributor (to the extent that is possible):
`tips.csv` - ID, dollar amount, date, anon contributor ID

I understand completely if this is too much to ask for. I would use this to see if developers respond differently when they depend on a few supporters giving a lot, or many giving little.

**Comment by chadwhitacre** [comment]
I think we can do that. Profile description would be something else that was public. I think that’s about it.

**Comment by chadwhitacre** [comment]
I guess some team info was also public.

**Comment by chadwhitacre** [comment]
Alright, started working on this. I got Gratipay running again locally (I had to upgrade psycopg2 to work around https://github.com/psycopg/psycopg2/issues/594), in order to remind myself what info we made public:

- The crucial fact is that we did not make individual gift amounts public, we only made the aggregate giving/receiving public.
- Another wrinkle is that we changed the model after the [Gratipocalypse](https://gratipay.news/gratipocalypse-42fd0ec0d9e8), inserting a "Projects" layer between givers and receivers and turning it more into a transaction than a gift.
- We also had the ability for folks to hide their giving/paying (but not their receiving/taking).

> I’m inclined to think a little more broadly and to provide two CSVs:

Sooooooo we can't do `tips.csv`, but we could add the following to `participants.csv`:

- `anonymous_giving`—true/false
- `giving_amount`—total dollar amount given on Gittip ~= Gratipay 1.0
- `giving_n`—total number of receivers the participant gave money to
- `giving_nweeks`—total number of weeks between first and last gift given
- `receiving_amount`—total dollar amount received on Gittip ~= Gratipay 1.0
- `receiving_n`—total number of givers the participant received money from
- `receiving_nweeks`—total number of weeks between first and last gift received
- `paid_amount`—total dollar amount paid on Gratipay 2.0
- `paid_n`—total number of projects the participant made payments to
- `paid_nweeks`—total number of weeks between first and last payment made
- `taken_amount`—total dollar amount taken on Gratipay 2.0
- `taken_n`—total number of projects the participant took money from
- `taken_nweeks`—total number of weeks between first and last payments taken

I suppose the giving/paying-related fields would be empty where `anonymous_giving` is true.

Would this still be interesting to you, @johanneswachs?

**Comment by johanneswachs** [comment]
Thank you for digging into the data @chadwhitacre. 

This would still be interesting, but a bit less so if we cannot say how much money individuals were receiving over time. Would it be possible to share how much money a developer received monthly or quarterly? I think this would go a long way to preserving privacy. Overall I am more interested in which developers received money, and roughly when, than who gave it.

**Comment by chadwhitacre** [comment]
@johanneswachs Most of the data listed above is stored directly on the `participants` table, so it would be quite easy to export from the last database snapshot (`2017-12-29`). The one thing not directly stored are the `nweeks` fields, but those wouldn't be too hard to compute from the `tips` and `takes` tables. You could compute an average but not much else.

It will take a fair bit more work to generate a more detailed time series. I think the best approach will be to work from the weekly database backups that I think I still have, loading each backup and then querying the computed values in `participants` to generate the above data for each participant for each week of Gittip/Gratipay. We ran on a weekly cycle, and the above data is what was public throughout, so anyone could've screen-scraped all along to come up with this data on their own. By working with all of the database backups rather than trying to work backwards using only the last snapshot, we can ensure that we deliver exactly what was public at any point in time (e.g., we'll be able to track `anonymous_giving` over time).

Does this seem reasonable?

If so, the next question is: do you care about payments made after the [Gratipocalypse](https://gratipay.news/gratipocalypse-42fd0ec0d9e8), or only about gifts given prior?

![Screen Shot 2019-10-27 at 7 32 02 PM](https://user-images.githubusercontent.com/134455/67643692-7af18980-f8f0-11e9-8a37-f5047f5b2af5.png)

As you can see, most of the interesting volume on Gratipay was prior to the change in our model. It will be easier to generate data and the resulting dataset will be simpler if we limit to Gittip/Gratipay 1.0.

**Comment by johanneswachs** [comment]
@chadwhitacre That is very reasonable. Thank you for proposing to go through all of this effort.

We can be happy without the post Gratipocalypse data, especially if it simplifies the extraction.

**Comment by Changaco** [comment]
Going through weekly backups isn't possible for Liberapay since we no longer have such backups (deleted them for GDPR compliance). Thankfully it's not necessary.

Below are DB queries to export aggregated weekly income data from Liberapay and Gratipay, and below that is a sample of the data from Liberapay.

---

First, a mapping of Liberapay user IDs to GitHub user IDs:

```sql
COPY (
    SELECT p.id as liberapay_user_id, e.user_id as github_user_id
      FROM participants p
      JOIN elsewhere e ON e.participant = p.id AND e.platform = 'github'
     WHERE NOT p.hide_receiving
) TO stdout (format csv, header true);
```

Next, the weekly data, including both the received amount and the number of donors:

```sql
CREATE FUNCTION currency_basket_to_text_for_csv(currency_basket)
RETURNS text AS $$
    DECLARE
        r text = ''; currency text; amount text;
    BEGIN
        FOR currency, amount IN SELECT * FROM jsonb_each_text($1.amounts) LOOP
            IF (length(r) > 0) THEN
                r := r || '+';
            END IF;
            r := r || currency || amount;
        END LOOP;
        RETURN r;
    END;
$$ LANGUAGE plpgsql IMMUTABLE STRICT;

COPY (
    SELECT tr.payday, tr.tippee AS recipient_liberapay_id, tr.npatrons, tr.income
      FROM ( SELECT payday.ts_start::date AS payday
                  , tr.tippee
                  , count(DISTINCT tr.tipper) AS npatrons
                  , currency_basket_to_text_for_csv(basket_sum(tr.amount)) AS income
               FROM transfers tr
               JOIN LATERAL (
                        SELECT payday.*
                          FROM paydays payday
                         WHERE payday.ts_end > tr.timestamp
                      ORDER BY payday.ts_end
                         LIMIT 1
                    ) payday ON true
              WHERE tr.context IN ('tip', 'take', 'final-gift')
           GROUP BY payday.ts_start::date, tr.tippee
           ) tr
      JOIN participants p ON p.id = tr.tippee
     WHERE NOT p.hide_receiving
  ORDER BY tr.payday, tr.tippee
) TO stdout (format csv, header true);

DROP FUNCTION currency_basket_to_text_for_csv(currency_basket);
```

Now the equivalent for Gratipay. First the mapping of Gratipay usernames to GitHub user IDs:

```sql
COPY (
    SELECT p.username as gratipay_username, e.user_id as github_user_id
      FROM participants p
      JOIN elsewhere e ON e.participant = p.username AND e.platform = 'github'
     WHERE NOT p.anonymous_receiving
) TO stdout (format csv, header true);
```

then the weekly data:

```sql
COPY (
    SELECT tr.payday, tr.tippee AS recipient_gratipay_username, tr.npatrons, tr.income
      FROM ( SELECT payday.ts_start::date AS payday
                  , tr.tippee
                  , count(DISTINCT tr.tipper) AS npatrons
                  , sum(tr.amount) AS income
               FROM transfers tr
               JOIN LATERAL (
                        SELECT payday.*
                          FROM paydays payday
                         WHERE payday.ts_end > tr.timestamp
                      ORDER BY payday.ts_end
                         LIMIT 1
                    ) payday ON true
              WHERE tr.context <> 'take-over'
           GROUP BY payday.ts_start::date, tr.tippee
           ) tr
      JOIN participants p ON p.username = tr.tippee
     WHERE NOT p.anonymous_receiving
  ORDER BY tr.payday, tr.tippee
) TO stdout (format csv, header true);
```

---

Sample of the weekly data from Liberapay:

```csv
payday,recipient_liberapay_id,npatrons,income
2016-02-10,1,3,EUR1.00
2016-02-10,2,1,EUR0.01
2016-02-10,12,2,EUR1.00
2016-02-10,69,2,EUR2.25
2016-02-10,83,2,EUR0.65
2016-02-10,119,1,EUR0.25
...
2019-10-30,1,446,AUD0.40+CAD2.26+CHF0.34+CZK13.00+DKK1.90+EUR172.63+GBP1.60+JPY32+NOK1.15+NZD0.78+PLN4.96+RUB21.53+SEK8.80+USD29.42
2019-10-30,12,50,EUR11.52+GBP0.34+NZD0.40+USD89.93
2019-10-30,65,1,EUR0.04
2019-10-30,69,2,EUR2.23
2019-10-30,119,10,EUR2.72
2019-10-30,133,2,EUR1.25
2019-10-30,136,4,EUR0.74+USD0.01
2019-10-30,185,2,EUR1.21
2019-10-30,288,2,EUR0.62
2019-10-30,325,10,EUR4.35
...
2019-10-30,1609148,8,USD43.28
2019-10-30,1609409,1,EUR0.86
2019-10-30,1609639,9,CAD1.00+EUR6.75+USD1.05
2019-10-30,1609879,2,USD7.30
2019-10-30,1610863,1,EUR1.00
```

---

I can't extract the Gratipay data myself because I only have the final DB backup (`2017-12-29.psql`) which doesn't contain the `anonymous_receiving` column (it was dropped in July 2017).

**Comment by chadwhitacre** [comment]
Nice work as usual, @Changaco.

> GDPR compliance

Do we have a GDPR issue here as well? At the very least I'm reluctant to publish whatever data files we do come up with. I think to avoid GDRP/bureaucratic overhead we should deliver them privately to @johanneswachs. If someone else wants to request these files as well we can decide then.

@johanneswachs Presumably you will be contacting former Gittip/Gratipay users based on the data we deliver to you, yes?

It's possible we could get some backlash from folks who wouldn't have wanted us to share their data. We are comfortable taking that risk, yes? I'm pretty sure I am, because I see this as a way for Gratipay's usefulness to extend beyond the end of the service itself. I think that benefit outweighs the risk.

**Comment by Changaco** [comment]
The DB queries only export data which is or was public, and I don't think providing that data in bulk creates any new risk, so I'm okay with publishing the files.

The users have or had a way to opt out of sharing this information, and we're honoring this privacy setting by filtering out the data based on the `anonymous_receiving` / `hide_receiving` column.

The CSV files produced by my queries can't be used to de-anonymize donors since they don't contain any data on donors or individual transfers, only aggregate weekly income amounts.

**Comment by chadwhitacre** [comment]
> the `anonymous_receiving` column (it was dropped in July 2017).

https://github.com/gratipay/gratipay.com/pull/4546 ftr

**Comment by chadwhitacre** [comment]
Actually, wait ... https://github.com/gratipay/gratipay.com/pull/3721? 🤔

**Comment by johanneswachs** [comment]
> 
> @johanneswachs Presumably you will be contacting former Gittip/Gratipay users based on the data we deliver to you, yes?
> 

This is a very good question. If you believe that this is necessary to remain GDPR complaint then I will obtain approval from individuals before proceeding. Could you estimate how many individuals received funds? In the meantime I will do some more research about GDPR.

**Comment by chadwhitacre** [comment]
@johanneswachs Sorry, I was actually assuming that you would need to contact devs in order to carry out your research. Is that not the case? Are you able to do your research without contacting individuals?

**Comment by johanneswachs** [comment]
@chadwhitacre I did some research and I believe that this use case falls under the research exemption (see: https://researchsupport.admin.ox.ac.uk/policy/data/exemptions#collapse470196). My impression is that if users gave Gratipay permission to record this data while the site was active, that data can now be used for "secondary" research purposes.

If you disagree or it makes you feel more comfortable, I can also contact the users and obtain permission. Would you mind telling me how many unique users received more than, say 50 dollars, from Gratipay? More than 500? Thank you!

**Comment by chadwhitacre** [comment]
Thanks for digging into that, @johanneswachs. I'm okay to proceed.

Scoping only to Gittip/Gratipay 1.0 tips, I count 442 users with GitHub accounts who received $50 or more cumulatively. I count 1,013 who received $10 or more.

```sql
  SELECT tippee, e.participant, total
    FROM (
        SELECT tippee, sum(amount) total
          FROM tips
      GROUP BY tippee
    ) AS t
    JOIN elsewhere e ON e.participant = t.tippee AND e.platform = 'github'
   WHERE total >= 10
ORDER BY total desc;
```

**Comment by Changaco** [comment]
@chadwhitacre The query should be on the `transfers` table, not `tips`. I see 1602 GitHub users who received $10 or more, and 934 who received $50 or more.

**Comment by chadwhitacre** [comment]
Right, sorry. I count 1,610 >= $10 and 937 >= $50. 🤔

```sql
  SELECT tippee, e.participant, total
    FROM (
        SELECT tippee, sum(amount) total
          FROM transfers
      GROUP BY tippee
    ) AS t
    JOIN elsewhere e
      ON e.participant = t.tippee AND e.platform = 'github'
   WHERE total >= 50
ORDER BY total desc;
```

**Comment by chadwhitacre** [comment]
Looks like I have an old db snapshot loaded. Maybe some refunds or something not accounted for here?

```
gratipay-bak=# select * from paydays order by ts_end desc limit 1;
┌─────┬───────────────────────────────┬───────────────────────────────┬────────┬────────┬───────┬────────┐
│ id  │           ts_start            │            ts_end             │ volume │ nusers │ stage │ nteams │
├─────┼───────────────────────────────┼───────────────────────────────┼────────┼────────┼───────┼────────┤
│ 329 │ 2017-06-15 10:01:17.707312-04 │ 2017-06-15 10:02:52.097269-04 │ 691.95 │    596 │     2 │    236 │
└─────┴───────────────────────────────┴───────────────────────────────┴────────┴────────┴───────┴────────┘
(1 row)

gratipay-bak=#
```

**Comment by chadwhitacre** [comment]
I've been unraveling the story of `anonymous_receiving` ...

- Requested in https://github.com/gratipay/gratipay.com/issues/1683.
- Implemented in https://github.com/gratipay/gratipay.com/pull/1847.
  - Redux of https://github.com/gratipay/gratipay.com/pull/1721.
- Removed in https://github.com/gratipay/gratipay.com/pull/3721, because with Gratipay 2.0 we went back to public receiving/taking.

That means we need to go back to a snapshot prior to https://github.com/gratipay/gratipay.com/pull/3721 being deployed in order to read `anonymous_receiving`. I'm still leaning towards iterating over all of the db snapshots so that we can observe changes to `anonymous_receiving` over time as opposed to simply taking the last value before the Gratipocalypse.

**Comment by Changaco** [comment]
`anonymous_receiving` was always retroactive, turning it on would hide data that was previously public, and turning it back off would republish everything.

---

> Right, sorry. I count 1,610 >= $10 and 937 >= $50. :thinking:

I'm filtering out transfers between accounts belonging to the same person.

```sql
SELECT count(*) FILTER (WHERE total >= 50), count(*) FILTER (WHERE total >= 10), count(*)
  FROM ( SELECT tippee, sum(amount) AS total
           FROM transfers
          WHERE context <> 'take-over'
       GROUP BY tippee
       ) t
  JOIN elsewhere e ON e.participant = t.tippee AND e.platform = 'github'
```

**Comment by chadwhitacre** [comment]
Nice. TIL `FILTER`. I get 1603 and 933 on this 2017 backup ... only off by one. :)

I'm going to load a backup from 2015 so I can get `anonymous_receiving` ...

**Comment by chadwhitacre** [comment]
The Gratipocalypse was 2015-05-07, so I loaded the snapshot from 2015-05-06.

I want to aggregate by month, because a) weekly was always weird, b) @johanneswachs [asked](https://github.com/gratipay/inside.gratipay.com/issues/1227#issuecomment-546238825) for monthly, and c) it preserves privacy a little more.

Here is the query I've come up with, based on @Changaco's work:

```sql
    SELECT substr(tr.month::text, 0, 8) AS month
         , e.user_name                  AS username
         , e.user_id                    AS id
         , tr.amount                    AS "sum"
         , tr.n
      FROM ( SELECT date_trunc('month', payday.ts_start) AS month
                  , tr.tippee
                  , count(DISTINCT tr.tipper) AS n
                  , sum(tr.amount) AS amount
               FROM transfers tr
               JOIN LATERAL (
                        SELECT payday.*
                          FROM paydays payday
                         WHERE payday.ts_end > tr.timestamp
                      ORDER BY payday.ts_end
                         LIMIT 1
                    ) payday ON true
              WHERE tr.context <> 'take-over'
           GROUP BY date_trunc('month', payday.ts_start), tr.tippee
           ) tr
      JOIN participants p ON p.username = tr.tippee
      JOIN elsewhere e ON e.participant = p.username AND e.platform = 'github'
     WHERE NOT p.anonymous_receiving
  ORDER BY tr.month, "sum" desc
```

The output is monthly receiving for GitHub users; here is a snippet:

```
┌─────────┬─────────────────────────┬──────────┬─────────┬─────┐
│  month  │        username         │    id    │   sum   │  n  │
├─────────┼─────────────────────────┼──────────┼─────────┼─────┤
│ 2012-06 │ whit537                 │ 134455   │   19.63 │  15 │
│ 2012-06 │ antirez                 │ 65632    │   11.77 │  16 │
│ 2012-06 │ kennethreitz            │ 119893   │   10.08 │  20 │
│ 2012-06 │ steveklabnik            │ 27786    │    9.92 │   9 │
│ 2012-06 │ alex                    │ 772      │    8.72 │  17 │
│ 2012-06 │ wycats                  │ 4        │    8.48 │   7 │
│ 2012-06 │ tenderlove              │ 3124     │    8.24 │   6 │
│ 2012-06 │ jezdez                  │ 1610     │    7.20 │  15 │
│ 2012-06 │ mitsuhiko               │ 7396     │    5.12 │  14 │
│ 2012-06 │ jnoller                 │ 51528    │    4.96 │   7 │
│ 2012-06 │ ericholscher            │ 25510    │    4.72 │   9 │
│ 2012-06 │ jeresig                 │ 1615     │    4.48 │  16 │
...
```

Will that give you what you need for your research, @johanneswachs? If so I will export to CSV and post it here.

**Comment by johanneswachs** [comment]
@chadwhitacre absolutely - this looks perfect!

**Comment by chadwhitacre** [comment]
Here you go, @johanneswachs, et al.:

### [dollars-per-month-to-github-users.csv.gz](https://github.com/gratipay/inside.gratipay.com/files/3846133/dollars-per-month-to-github-users.csv.gz)

(I gzipped it because GitHub doesn't accept `.csv` for file uploads. I also considered a gist but the file is too large for GitHub to make it "beautiful and searchable," soooo ... meh.)

I'll leave this ticket open for a week or so in case you have any questions. I would love to find out if and when this dataset ever makes it into a publication. I'll probably have re-archived this repo by then, so could you let me know via [email](mailto:chad@zetaweb.com)? Thanks and good luck! :)

### Fields

| name | description |
|:---------|:--------|
| `month` | in `YYYY-MM` format, e.g., `2012-06`; Gittip/Gratipay 1.0 operated for 35 months (a month short of three years!), from June, 2012 through April, 2015 |
| `sum` | the total amount, in USD ($), that the GitHub user received during the month via Gittip/Gratipay |
| `n` | the number of payments to the GitHub user during the month; payments were on a weekly cycle, and are aggregated here |
| `github_user_id` | the GitHub user's immutable ID |
| `github_username_in_2015` | the GitHub user's username; this is mutable, and a number of these will indeed have mutated since we first recorded a value ;) |

### SQL

```sql
COPY (
    SELECT substr(tr.month::text, 0, 8) AS month
         , tr.amount                    AS "sum"
         , tr.n
         , e.user_id                    AS github_user_id
         , e.user_name                  AS github_username_in_2015
      FROM ( SELECT date_trunc('month', payday.ts_start) AS month
                  , tr.tippee
                  , count(DISTINCT tr.tipper) AS n
                  , sum(tr.amount) AS amount
               FROM transfers tr
               JOIN LATERAL (
                        SELECT payday.*
                          FROM paydays payday
                         WHERE payday.ts_end > tr.timestamp
                      ORDER BY payday.ts_end
                         LIMIT 1
                    ) payday ON true
              WHERE tr.context <> 'take-over'
           GROUP BY date_trunc('month', payday.ts_start), tr.tippee
           ) tr
      JOIN participants p ON p.username = tr.tippee
      JOIN elsewhere e ON e.participant = p.username AND e.platform = 'github'
     WHERE NOT p.anonymous_receiving
  ORDER BY tr.month, "sum" desc
) TO '/Users/whit537/workbench/gratipay/logs/misc/1227/dollars-per-month-to-github-users.csv'
  (format csv, header true);
```

**Comment by johanneswachs** [comment]
@chadwhitacre - I will be sure to let you know! Thank you very much again for digging back into the past for this. I am excited to investigate the data.
